# 80 proj

**Name(s)**: Adrian Apsay and Ethan Flores

**Website Link**: https://adrianapsay.github.io/power_outage/

In [79]:
import pandas as pd
import numpy as np
from pathlib import Path

import plotly.express as px
pd.options.plotting.backend = 'plotly'

from dsc80_utils import * # Feel free to uncomment and use this.
import openpyxl

# for modeling
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import precision_score

In [80]:
# from lec02 to display all of the df
from IPython.display import display

from IPython.display import display
def display_df(df, rows=5, cols=None):
    """Displays n rows and cols from df."""
    with pd.option_context("display.max_rows", rows,
                           "display.max_columns", cols):
        display(df)

## Step 1: Introduction

In [81]:
# TODO

## Step 2: Data Cleaning and Exploratory Data Analysis

In [82]:
df = pd.read_excel('outage.xlsx')
df

Major power outage events in the continental U.S. Unnamed: 1 Unnamed: 2  \
0                 Time period: January 2000 - July 2016        NaN        NaN   
1     Regions affected: Outages reported in this dat...        NaN        NaN   
2                                                   NaN        NaN        NaN   
...                                                 ...        ...        ...   
1537                                                NaN       1532       2009   
1538                                                NaN       1533       2009   
1539                                                NaN       1534       2000   

     Unnamed: 3  ... Unnamed: 53 Unnamed: 54 Unnamed: 55 Unnamed: 56  
0           NaN  ...         NaN         NaN         NaN         NaN  
1           NaN  ...         NaN         NaN         NaN         NaN  
2           NaN  ...         NaN         NaN         NaN         NaN  
...         ...  ...         ...         ...         ...         ...  
1537          8  ...        0.15       98.31        1.69        1.69  
1538          8  ...        0.15       98.31        1.69        1.69  
1539        NaN  ...        0.02       85.76       14.24         2.9  

[1540 rows x 57 columns]

In [83]:
df.columns = list(df.iloc[4])
df = df.drop(range(0,5))
df = df.reset_index().iloc[1:,2:]
df = df.reset_index().drop(columns='index')
df

OBS  YEAR MONTH    U.S._STATE  ... AREAPCT_UC PCT_LAND PCT_WATER_TOT  \
0        1  2011     7     Minnesota  ...        0.6    91.59          8.41   
1        2  2014     5     Minnesota  ...        0.6    91.59          8.41   
2        3  2010    10     Minnesota  ...        0.6    91.59          8.41   
...    ...   ...   ...           ...  ...        ...      ...           ...   
1531  1532  2009     8  South Dakota  ...       0.15    98.31          1.69   
1532  1533  2009     8  South Dakota  ...       0.15    98.31          1.69   
1533  1534  2000   NaN        Alaska  ...       0.02    85.76         14.24   

     PCT_WATER_INLAND  
0                5.48  
1                5.48  
2                5.48  
...               ...  
1531             1.69  
1532             1.69  
1533              2.9  

[1534 rows x 56 columns]

In [84]:
df.sort_values(by='YEAR')

OBS  YEAR MONTH      U.S._STATE  ... AREAPCT_UC PCT_LAND PCT_WATER_TOT  \
766  767  2000   NaN  North Carolina  ...       2.11    90.34          9.66   
235  236  2000     3           Texas  ...       0.58    97.26          2.74   
239  240  2000   NaN           Texas  ...       0.58    97.26          2.74   
..   ...   ...   ...             ...  ...        ...      ...           ...   
657  658  2016     5            Utah  ...        0.2    96.79          3.21   
655  656  2016     1            Utah  ...        0.2    96.79          3.21   
683  684  2016     2            Utah  ...        0.2    96.79          3.21   

    PCT_WATER_INLAND  
766             7.53  
235             2.09  
239             2.09  
..               ...  
657             3.21  
655             3.21  
683             3.21  

[1534 rows x 56 columns]

In [85]:
df.columns

Index(['OBS', 'YEAR', 'MONTH', 'U.S._STATE', 'POSTAL.CODE', 'NERC.REGION',
       'CLIMATE.REGION', 'ANOMALY.LEVEL', 'CLIMATE.CATEGORY',
       'OUTAGE.START.DATE', 'OUTAGE.START.TIME', 'OUTAGE.RESTORATION.DATE',
       'OUTAGE.RESTORATION.TIME', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL',
       'HURRICANE.NAMES', 'OUTAGE.DURATION', 'DEMAND.LOSS.MW',
       'CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE', 'IND.PRICE',
       'TOTAL.PRICE', 'RES.SALES', 'COM.SALES', 'IND.SALES', 'TOTAL.SALES',
       'RES.PERCEN', 'COM.PERCEN', 'IND.PERCEN', 'RES.CUSTOMERS',
       'COM.CUSTOMERS', 'IND.CUSTOMERS', 'TOTAL.CUSTOMERS', 'RES.CUST.PCT',
       'COM.CUST.PCT', 'IND.CUST.PCT', 'PC.REALGSP.STATE', 'PC.REALGSP.USA',
       'PC.REALGSP.REL', 'PC.REALGSP.CHANGE', 'UTIL.REALGSP', 'TOTAL.REALGSP',
       'UTIL.CONTRI', 'PI.UTIL.OFUSA', 'POPULATION', 'POPPCT_URBAN',
       'POPPCT_UC', 'POPDEN_URBAN', 'POPDEN_UC', 'POPDEN_RURAL',
       'AREAPCT_URBAN', 'AREAPCT_UC', 'PCT_LAND', 'PCT_WATER_TOT',
    

In [86]:
df[(df['U.S._STATE'] == 'California')]

OBS  YEAR MONTH  U.S._STATE  ... AREAPCT_UC PCT_LAND PCT_WATER_TOT  \
1075  1076  2007     9  California  ...       0.59    95.16          4.84   
1076  1077  2008     5  California  ...       0.59    95.16          4.84   
1077  1078  2006     5  California  ...       0.59    95.16          4.84   
...    ...   ...   ...         ...  ...        ...      ...           ...   
1282  1283  2003    12  California  ...       0.59    95.16          4.84   
1283  1284  2008    11  California  ...       0.59    95.16          4.84   
1284  1285  2014     4  California  ...       0.59    95.16          4.84   

     PCT_WATER_INLAND  
1075             1.73  
1076             1.73  
1077             1.73  
...               ...  
1282             1.73  
1283             1.73  
1284             1.73  

[210 rows x 56 columns]

### **exploring the relationship between duration/severity of power outages based on the economy of each state**

We want to investigate the relationship of outage duration and severity based on the economy of a given state. We hypothesize that states with lower economic productivity results in greater outage duration and severity due to the lack of resources and funds to maintain infrastructure that powers the electricity of the state. This question will provide insight into the economic productivity and performance of each state, making it useful for comparing the economic status of different states based on the impact of power outages. By exploring the relationship between outage duration and severity with the economic output of a state along with their utility contribution to the total GSP per-state, as represented by PC.REALGSP.STATE and UTIL.CONTRI respectively, we can assess how disruptions in electricity supply that provide power is affected by the economic well-being of each state. This analysis can help identify potential vulnerabilities in states with lower measures of economic productivity per capita and inform policy decisions aimed at improving resilience to power outages as well as enhancing overall economic productivity to fund infrastructure that fuels this power.

The Power Outages dataset is an Excel file that also had descriptions for each column. Reading into the Excel file locally in order to perform cleaning and EDA-related tasks displayed a deformed dataframe with columns that aren't necessary to the relationship we wanted to investigate. As such, we decided to keep only YEAR, U.S._STATE, OUTAGE.DURATION, PC.REALGSP.STATE, UTIL.REALGSP, UTIL.CONTRI. These columns will be particularly useful for our exploratory data analysis as they provide essential information that allows us to investigate the relationship between power outage characteristics and economic factors at the state level.

In [87]:
df.columns

Index(['OBS', 'YEAR', 'MONTH', 'U.S._STATE', 'POSTAL.CODE', 'NERC.REGION',
       'CLIMATE.REGION', 'ANOMALY.LEVEL', 'CLIMATE.CATEGORY',
       'OUTAGE.START.DATE', 'OUTAGE.START.TIME', 'OUTAGE.RESTORATION.DATE',
       'OUTAGE.RESTORATION.TIME', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL',
       'HURRICANE.NAMES', 'OUTAGE.DURATION', 'DEMAND.LOSS.MW',
       'CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE', 'IND.PRICE',
       'TOTAL.PRICE', 'RES.SALES', 'COM.SALES', 'IND.SALES', 'TOTAL.SALES',
       'RES.PERCEN', 'COM.PERCEN', 'IND.PERCEN', 'RES.CUSTOMERS',
       'COM.CUSTOMERS', 'IND.CUSTOMERS', 'TOTAL.CUSTOMERS', 'RES.CUST.PCT',
       'COM.CUST.PCT', 'IND.CUST.PCT', 'PC.REALGSP.STATE', 'PC.REALGSP.USA',
       'PC.REALGSP.REL', 'PC.REALGSP.CHANGE', 'UTIL.REALGSP', 'TOTAL.REALGSP',
       'UTIL.CONTRI', 'PI.UTIL.OFUSA', 'POPULATION', 'POPPCT_URBAN',
       'POPPCT_UC', 'POPDEN_URBAN', 'POPDEN_UC', 'POPDEN_RURAL',
       'AREAPCT_URBAN', 'AREAPCT_UC', 'PCT_LAND', 'PCT_WATER_TOT',
    

Kept Columns: YEAR, U.S._STATE, OUTAGE.DURATION, UTIL.REALGSP, UTIL.CONTRI
We kept these columns as they are useful for exploratory data analysis (EDA) because they provide essential information that allows us to investigate the relationship between power outage characteristics and economic factors at the state level overtime each year.

By analyzing these columns together, we can investigate how power outage duration and severity vary across different 
states and how they may be influenced by economic factors such as state-level economic productivity and their contribution in the utility sector. This analysis will highlight insights in areas where power outages are more rampant and/or more severe conditioned on the economic productivity and utility contribution for each state. 



Some columns we want to explain even further and our choices into keeping these columns are UTIL.REALGSP and UTIL.CONTRI:

Real Gross State Product of Utility Sectory by Each State (UTIL.REALGSP) will provide insights into the economic output of each state generated by the utility sector, helping to understand the importance of utilities in the state's economic activity and how different patterns may arise in outage duration and severity based on this metric.

Utility Industry's Contribution to Real Gross State Product (UTIL.CONTRI) represents the contribution of utility services to the overall economy of the state, that is, the share of the state's economy that may be contributed by utility services. It highlights the importance of utilities within each state's economy, which could be useful in understanding patterns in power outages (i.e. do states with more concentration on utilities in their economy have lesser outage duration and/or severity?)

In [88]:
concise_df = df[["YEAR", "U.S._STATE", "OUTAGE.DURATION", "UTIL.REALGSP", "UTIL.CONTRI"]]
concise_df

YEAR    U.S._STATE OUTAGE.DURATION UTIL.REALGSP UTIL.CONTRI
0     2011     Minnesota            3060         4802        1.75
1     2014     Minnesota               1         5226        1.79
2     2010     Minnesota            3000         4571        1.71
...    ...           ...             ...          ...         ...
1531  2009  South Dakota              59          606        1.66
1532  2009  South Dakota             181          606        1.66
1533  2000        Alaska             NaN          724        2.01

[1534 rows x 5 columns]

For almost the entirety of our EDA, we will be creating a dataframe that offers more nuanced information to garner deeper insights into the relationships between power outages and economic factors. That is, each row represents a power outage at a certain year and state (YEAR and U.S._STATE are also columns). We've also added additional columns:
1. `Total_Outages_Per_State_That_Year`: the total number of power outages per state that year
2. `Avg_Outage_Duration_Per_State_That_Year`: the average duration of power outages per state and year
3. `Outage_Severity`: interpreted as a measure of the severity of power outages in each state that year, taking into account both total number of outages and average duration for that state per year. This was computed by multiplying the
total number of outages by the average outage duration (`Total_Outages_Per_State_That_Year * Avg_Outage_Duration_Per_State_That_Year`)
4. `UTIL.REALGSP`: From original dataset (explained why this column was kept prior)
5. `UTIL.CONTRI`: From original dataset (explained why this column was kept prior)

In [89]:
# Calculate the total number of power outages per state and year
outages_per_state_year = df.groupby(['YEAR', 'U.S._STATE']).size().reset_index(name='Total_Outages_Per_State_That_Year')

# Calculate the average duration of power outages per state and year
avg_duration_per_state_year = df.groupby(['YEAR', 'U.S._STATE'])['OUTAGE.DURATION'].mean().reset_index(name='Avg_Outage_Duration_Per_State_That_Year')

# Merge the statistics with the UTIL.REALGSP and UTIL.CONTRI data
gsp_data = df[['YEAR', 'U.S._STATE', 'UTIL.REALGSP', 'UTIL.CONTRI']].drop_duplicates()

# Merge the dataframes
merged_data = pd.merge(outages_per_state_year, avg_duration_per_state_year, on=['YEAR', 'U.S._STATE'])
merged_data = pd.merge(merged_data, gsp_data, on=['YEAR', 'U.S._STATE'])
merged_data

YEAR  U.S._STATE  Total_Outages_Per_State_That_Year  \
0    2000     Alabama                                  3   
1    2000      Alaska                                  1   
2    2000     Arizona                                  2   
..    ...         ...                                ...   
402  2016        Utah                                  5   
403  2016  Washington                                  7   
404  2016   Wisconsin                                  1   

     Avg_Outage_Duration_Per_State_That_Year UTIL.REALGSP UTIL.CONTRI  
0                                    2247.00         5704         3.8  
1                                        NaN          724        2.01  
2                                      66.00         5745        2.84  
..                                       ...          ...         ...  
402                                    72.00         1576        1.15  
403                                   744.33         3504        0.83  
404                                  1605.00         4900        1.77  

[405 rows x 6 columns]

In [90]:
# interpreted as a measure of the severity of power outages in each state, taking into account both frequency and duration
merged_data['Outage_Severity'] = merged_data['Total_Outages_Per_State_That_Year'] * merged_data['Avg_Outage_Duration_Per_State_That_Year']
merged_data

YEAR  U.S._STATE  Total_Outages_Per_State_That_Year  \
0    2000     Alabama                                  3   
1    2000      Alaska                                  1   
2    2000     Arizona                                  2   
..    ...         ...                                ...   
402  2016        Utah                                  5   
403  2016  Washington                                  7   
404  2016   Wisconsin                                  1   

     Avg_Outage_Duration_Per_State_That_Year UTIL.REALGSP UTIL.CONTRI  \
0                                    2247.00         5704         3.8   
1                                        NaN          724        2.01   
2                                      66.00         5745        2.84   
..                                       ...          ...         ...   
402                                    72.00         1576        1.15   
403                                   744.33         3504        0.83   
404                                  1605.00         4900        1.77   

     Outage_Severity  
0            6741.00  
1                NaN  
2             132.00  
..               ...  
402           360.00  
403          5210.33  
404          1605.00  

[405 rows x 7 columns]

Here are the first 5 rows of our new dataframe that we will be using for our Exploratory Data Analysis:

In [91]:
merged_data.head()

YEAR  U.S._STATE  Total_Outages_Per_State_That_Year  \
0  2000     Alabama                                  3   
1  2000      Alaska                                  1   
2  2000     Arizona                                  2   
3  2000  California                                  1   
4  2000    Delaware                                  1   

   Avg_Outage_Duration_Per_State_That_Year UTIL.REALGSP UTIL.CONTRI  \
0                                   2247.0         5704         3.8   
1                                      NaN          724        2.01   
2                                     66.0         5745        2.84   
3                                      NaN        29581        1.79   
4                                      NaN         1016        2.02   

   Outage_Severity  
0           6741.0  
1              NaN  
2            132.0  
3              NaN  
4              NaN

In [92]:
# merged_data[merged_data['U.S._STATE'] == "California"].sort_values(by='Total_Outages_Per_State_That_Year', ascending=False)


### Univariate Analysis

Plot 1: We'll take a look at the distribution of the `Avg_Outage_Duration_Per_State_That_Year` column to understand how outage durations vary across states per year.

In [93]:
import plotly.express as px

fig_total_outages = px.histogram(merged_data, 
                                 x='Avg_Outage_Duration_Per_State_That_Year',
                                 nbins=50, 
                                 title='Distribution of Average Outage Durations Per State That Year')
fig_total_outages.update_layout(xaxis_title='Average Outage Durations (minutes) Per State That Year', 
                                yaxis_title='Frequency')
fig_total_outages.show()

This histogram provides a visual understanding of how the average outage durations are distributed across different states and years, which can highlight trends or outliers worth investigating further. We can see that this histogram is right-skewed, where a majority of average outage durations are clustered towards the lower end of averages. Most states seem to experience relatively shorter average outage durations with the cut-off being around 15,000 minutes.

Plot 2: We will examine the distribution of the `Outage_Severity` column using a histogram in order to understand the severity of power outages across different states and years as a whole, which may serve as a better measurement in comparison to just total outages and average outage duration for each state per year since it accounts for both.

In [94]:
import plotly.express as px


fig = px.histogram(merged_data, x='Outage_Severity', title='Distribution of Power Outage Severity')
fig.update_layout(xaxis_title='Outage Severity (Min)', yaxis_title='Frequency')
fig.show()


This histogram plot visualizes the distribution of power outage severity across different states and years. The x-axis represents the severity of power outages, while the y-axis represents the frequency of occurrences. From the plot, we can observe the distribution of outage severity and identify any patterns or outliers present. In this case, we see that this histogram is right-skewed, with a majority of outage severities prevalent up to 50,000 minutes.

### Bivariate Analysis

Let's examine the relationship between `Outage_Severity` and `UTIL.REALGSP` using a scatter plot. This will help us understand if there's any association between the severity of power outages and the real gross state product contributed by the utility industry (economic output by the utility industry of the state) and further explore the relationship we're trying to investigate: the outage duration and severity based on the economy of a given state.

In [95]:
# Create a scatter plot for Outage_Severity vs UTIL.REALGSP
fig = px.scatter(merged_data, x='UTIL.REALGSP', y='Outage_Severity', color='U.S._STATE',
                 title='Relationship between Outage Severity and Utility Real GSP')
fig.update_layout(xaxis_title='Utility Real GSP', yaxis_title='Outage Severity')
fig.show()


#### Aggregated Analysis

In [96]:
# Calculate mean values for UTIL.REALGSP and UTIL.CONTRI
mean_realgsp = merged_data['UTIL.REALGSP'].mean()
mean_contrib = merged_data['UTIL.CONTRI'].mean()

# Assign high and low labels for UTIL.REALGSP and UTIL.CONTRI
merged_data['UTIL.REALGSP Label'] = merged_data['UTIL.REALGSP'].apply(lambda x: 'High' if x > mean_realgsp else 'Low')
merged_data['UTIL.CONTRI Label'] = merged_data['UTIL.CONTRI'].apply(lambda x: 'High' if x > mean_contrib else 'Low')

# Create pivot table
pivot_table = merged_data.pivot_table(values=['Total_Outages_Per_State_That_Year', 'Avg_Outage_Duration_Per_State_That_Year'], 
                                      index=['UTIL.REALGSP Label', 'UTIL.CONTRI Label'], 
                                      aggfunc={'Total_Outages_Per_State_That_Year': 'sum', 'Avg_Outage_Duration_Per_State_That_Year': 'mean'})
pivot_table


Avg_Outage_Duration_Per_State_That_Year  \
UTIL.REALGSP Label UTIL.CONTRI Label                                            
High               High                                               2954.52   
                   Low                                                3513.45   
Low                High                                               2852.20   
                   Low                                                1737.03   

                                      Total_Outages_Per_State_That_Year  
UTIL.REALGSP Label UTIL.CONTRI Label                                     
High               High                                             386  
                   Low                                              357  
Low                High                                             304  
                   Low                                              487

In [97]:
import plotly.express as px

pivot_table = pivot_table.reset_index()
# Plot average outage duration
fig_avg_duration = px.bar(pivot_table, 
                          x='UTIL.REALGSP Label', 
                          y='Avg_Outage_Duration_Per_State_That_Year', 
                          color='UTIL.CONTRI Label', 
                          barmode='group', 
                          title='Average Outage Duration by Real GSP and Utility Contribution Labels')
fig_avg_duration.update_layout(xaxis_title='UTIL.REALGSP Label', yaxis_title='Average Outage Duration (minutes)')
fig_avg_duration.show()

## Step 3: Assessment of Missingness

In [98]:
# TODO

## Step 4: Hypothesis Testing

In [99]:
# TODO

## Step 5: Framing a Prediction Problem

From our analysis we can see that there wasn't a prominent relationship between duration/severity of power outages and the economic characteristics of each state. As a result, we have shifted our focus into if we are able to predict the cause of an outage based on the duration of the outage and its economic characteristics for the given state.

We will employ a multiclass classification model through a Random Forest Classifier to accurately predict the labels in the `CAUSE.CATEGORY`. At this moment, we have only analyzed and placed an emphasis on the economic features of the dataset along with the duration of each outage. As a result, they will be our main features. They are as follows `CAUSE.CATEGORY`,`OUTAGE.DURATION`, `PC.REALGSP.STATE`, `PC.REALGSP.USA`,`PC.REALGSP.REL`,`PC.REALGSP.CHANGE`, `UTIL.REALGSP`, `TOTAL.REALGSP`, `UTIL.CONTRI`, `PI.UTIL.OFUSA`.

Due to the great imbalances in the `CAUSE.CATEGORY` column, we will opt for our test statistic to be F1-score.

In [100]:
# plotting value counts of the model to see class imbalances in cause category
df['CAUSE.CATEGORY'].dropna().value_counts()

severe weather                   763
intentional attack               418
system operability disruption    127
public appeal                     69
equipment failure                 60
fuel supply emergency             51
islanding                         46
Name: CAUSE.CATEGORY, dtype: int64

## Step 6: Baseline Model

In [101]:
# obtaining the economic features and outage duration to classify the cause categories
economic_features = ['PC.REALGSP.STATE', 'PC.REALGSP.USA', 'PC.REALGSP.REL', 
                            'PC.REALGSP.CHANGE', 'UTIL.REALGSP', 'TOTAL.REALGSP', 'UTIL.CONTRI', 'PI.UTIL.OFUSA']
features = ['CAUSE.CATEGORY', 'OUTAGE.DURATION'] + economic_features

# creating dataframe for the rest of the modeling section 
modeling_df = df[features].dropna()

# obtaining features and response variable
X = modeling_df.drop(columns=['CAUSE.CATEGORY'])
y = modeling_df['CAUSE.CATEGORY']

In [102]:
display_df(modeling_df.head())

CAUSE.CATEGORY OUTAGE.DURATION PC.REALGSP.STATE PC.REALGSP.USA  \
0      severe weather            3060            51268          47586   
1  intentional attack               1            53499          49091   
2      severe weather            3000            50447          47287   
3      severe weather            2550            51598          48156   
4      severe weather            1740            54431          49844   

  PC.REALGSP.REL PC.REALGSP.CHANGE UTIL.REALGSP TOTAL.REALGSP UTIL.CONTRI  \
0           1.08               1.6         4802        274182        1.75   
1           1.09               1.9         5226        291955        1.79   
2           1.07               2.7         4571        267895        1.71   
3           1.07               0.6         5364        277627        1.93   
4           1.09               1.7         4873        292023        1.67   

  PI.UTIL.OFUSA  
0           2.2  
1           2.2  
2           2.1  
3           2.2  
4           2.2

In [103]:
def evaluate_model(pipeline, index_name, X, y, grid_search = False):
    # data splitting into training and testing splits
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
    pipeline.fit(X_train, y_train)
    
    if grid_search:
        pipeline.fit(X_train, y_train)
        best_pipeline = pipeline.best_estimator_
        print("Best Parameters:", pipeline.best_params_)
    else:
        pipeline.fit(X_train, y_train)
        best_pipeline = pipeline
    
    # training accuracy
    y_train_pred = pipeline.predict(X_train)
    train_acc = accuracy_score(y_train, y_train_pred)
    
    # testing accuracy
    y_test_pred = pipeline.predict(X_test)
    test_acc = accuracy_score(y_test, y_test_pred)
    
    # classification metrics
    precision = precision_score(y_test, y_test_pred, average='weighted')
    recall = recall_score(y_test, y_test_pred, average='weighted')
    f1 = f1_score(y_test, y_test_pred, average='weighted')
    
    metrics_dict = {
        'Training Accuracy': [train_acc],
        'Testing Accuracy': [test_acc],
        'Precision': [precision],
        'Recall': [recall],
        'F1-Score': [f1]
    }
       
    return pd.DataFrame(metrics_dict, index = [index_name]), pipeline

In [104]:
# baseline model pipeline, no transforms as all of the data is numerical
baseline_pipeline = Pipeline([
#     ('preprocessor', preproc), 
    ('clf', RandomForestClassifier())
])

baseline_df = evaluate_model(baseline_pipeline, 'Random Forest, Baseline Model', X, y)[0]
baseline_df

Training Accuracy  Testing Accuracy  Precision  \
Random Forest, Baseline Model                1.0               0.7       0.67   

                               Recall  F1-Score  
Random Forest, Baseline Model     0.7      0.68

## Step 7: Final Model

For this section I am choosing to opt for a GridSearchCV ot find the hyperparameters of the my RandomForest. The hyperparameters I opted to choose from are:

- clf__n_estimators: We choose [50, 100] to balance model complexity and computational efficiency, ensuring an adequate number of trees for robustness without excessive runtime.

- clf__max_features: Options like 'auto' and 'sqrt' are chosen to vary the number of features considered at each split, accommodating different levels of feature selection and model complexity.

- clf__max_depth: [None, 10, 20] is selected to control tree depth, balancing model expressiveness with prevention of overfitting by limiting tree complexity.

- clf__min_samples_split: Options [2, 5] are set to impose a minimum number of samples required to split a node, promoting generalization and avoiding overly specific splits.

- clf__min_samples_leaf: [1, 2] is chosen to set a minimum number of samples required at each leaf node, aiming to prevent the model from fitting noise and improving generalization.

In [109]:
# creating final model, transforming 4 columns utilizing StandardScaler
final_preproc = ColumnTransformer(
    transformers=[
        ('std-scaler', StandardScaler(), ['PC.REALGSP.STATE', 'PC.REALGSP.USA', 'PC.REALGSP.REL', 'PC.REALGSP.CHANGE'])
    ],
    remainder='passthrough'
)

# Define final pipeline with preprocessing and RandomForestClassifier
final_pipeline = Pipeline([
    ('preprocessor', final_preproc),
    ('clf', RandomForestClassifier())
])

# parameter grid for GridSearchCV
param_grid = {
    'clf__n_estimators': [20, 50, 100],
    'clf__max_features': ['sqrt'],
    'clf__max_depth': [None, 20, 40],
    'clf__min_samples_split': [2, 5],
    'clf__min_samples_leaf': [1, 2],
}


# GridSearchCV on final_pipeline
grid_search = GridSearchCV(final_pipeline, param_grid=param_grid, cv=5, scoring='accuracy')

final_df, optimized_model = evaluate_model(grid_search, 'Random Forest, Transformers with GridSearch, Final Model', X, y, grid_search=True)
final_df = pd.concat([baseline_df, final_df])

final_df

Best Parameters: {'clf__max_depth': 20, 'clf__max_features': 'sqrt', 'clf__min_samples_leaf': 2, 'clf__min_samples_split': 2, 'clf__n_estimators': 50}


Training Accuracy  \
Random Forest, Baseline Model                                     1.0   
Random Forest, Transformers with GridSearch, Fi...                0.9   

                                                    Testing Accuracy  \
Random Forest, Baseline Model                                   0.70   
Random Forest, Transformers with GridSearch, Fi...              0.73   

                                                    Precision  Recall  \
Random Forest, Baseline Model                            0.67    0.70   
Random Forest, Transformers with GridSearch, Fi...       0.71    0.73   

                                                    F1-Score  
Random Forest, Baseline Model                           0.68  
Random Forest, Transformers with GridSearch, Fi...      0.71

## Step 8: Fairness Analysis

Null Hypothesis: Our model is fair. Its precision for severe weather and intentional attack are roughly the same, and any differences are due to random chance.

Alternate Hypothesis: Our model is unfair. Its precision for intentional attack is lower than its precision for severe health.

In [131]:
# splitting data for the permutation tests
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

# predicting labels for the test set using the optimized model
predictions = optimized_model.predict(X_test)
random_labels = np.random.choice(["severe weather", "intentional attack"], size=len(X_test))

# obtain precision scores for each label
prec_severe = precision_score(random_labels, predictions, labels=["severe weather"], average=None)[0]
prec_intentional = precision_score(random_labels, predictions, labels=["intentional attack"], average=None)[0]

# calculating observed difference in precision
observed_difference = prec_intentional - prec_severe

# perform random permutations for 5000 iterations
num_permutations = 5000
permuted_differences = []
for i in range(num_permutations):
    # shuffle predictions randomly
    np.random.shuffle(predictions)
    
    # compute precision scores for shuffled data
    shuffled_severe = precision_score(synthetic_ground_truth, predictions, labels=["severe weather"], average=None)[0]
    shuffled_intentional = precision_score(synthetic_ground_truth, predictions, labels=["intentional attack"], average=None)[0]
    
    # calculate difference in precision for shuffled data
    permuted_differences.append(shuffled_intentional - shuffled_severe)

# calculate p-value
p_value = (np.abs(permuted_differences) >= np.abs(observed_difference)).mean()
print(f"p-value: {p_value:.4f}")

p-value: 0.8878


The p-value of 0.8878 indicates insufficient evidence to reject the null hypothesis that the model's precision for severe weather and intentional attacks is not significantly different. This suggests that any observed variations in precision between the two categories are likely due to random chance rather than systematic bias. Therefore, the model appears to be fair in its performance across both types of events based on the available data